sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/AccessTokenCache.java

```diff
@@ -0,0 +1,165 @@
+
+package com.azure.core.experimental.credential;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoProcessor;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+public class AccessTokenCache {
+    // The delay after a refresh to attempt another token refresh
+    private static final Duration REFRESH_DELAY = Duration.ofSeconds(30);
+    // the offset before token expiry to attempt proactive token refresh
+    private static final Duration REFRESH_OFFSET = Duration.ofMinutes(5);
+    private final AtomicReference<MonoProcessor<AccessToken>> wip;
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
+    private final Supplier<Mono<AccessToken>> tokenSupplier;
+    private final Predicate<AccessToken> shouldRefresh;
+    private final ClientLogger logger = new ClientLogger(AccessTokenCache.class);
+
+    /**
+     * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
+     *
+     * @param tokenSupplier a method to get a new token
+     */
+    public AccessTokenCache(Supplier<Mono<AccessToken>> tokenSupplier) {
+        this.wip = new AtomicReference<>();
+        this.tokenSupplier = tokenSupplier;
+        this.shouldRefresh = accessToken -> OffsetDateTime.now()
+            .isAfter(accessToken.getExpiresAt().minus(REFRESH_OFFSET));
+    }
+
+    /**
+     * Asynchronously get a token from either the cache or replenish the cache with a new token.
+     * @return a Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> getToken() {
+        return getToken(this.tokenSupplier, false);
+    }
+
+    /**
+     * Asynchronously get a token from either the cache or replenish the cache with a new token.
+     *
+     * @param tokenSupplier The method to get a new token
+     * @param forceRefresh The flag indicating if the cache needs to be skipped and a token needs to be fetched via the
+     * credential.
+     * @return The Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> getToken(Supplier<Mono<AccessToken>> tokenSupplier, boolean forceRefresh) {
+        return Mono.defer(() -> {
+            try {
+                if (wip.compareAndSet(null, MonoProcessor.create())) {
+                    final MonoProcessor<AccessToken> monoProcessor = wip.get();
+                    OffsetDateTime now = OffsetDateTime.now();
+                    Mono<AccessToken> tokenRefresh;
+                    Mono<AccessToken> fallback;
+                    if (forceRefresh) {
+                        tokenRefresh = Mono.defer(tokenSupplier);
```

Comment:
Any reason for having nested `defer()`? This gets executed only if the outer `Mono` is subscribed to.

Reply:
yes, this should get executed only on subscription downstream when chained with materialize.